### ISSUE TITLE
OSX tests that spawn windows blast the WindowServer process

### ISSUE THREAD

**Issue by larsbergstrom** [servo/servo#9513]
On our OSX builders, I've noticed that while we are running at least test-css, the act of spawning the window and shutting it down while running tests causes the WindowServer process to basically eat up an entire CPU. We could dramatically reduce our runtime by fixing this.

cc @metajack

**Comment by larsbergstrom** [comment]
I'm going to look at glutin to see if it already supports headless rendering on OSX, and if not, play around with http://renderingpipeline.com/2012/05/windowless-opengl-on-macos-x/

cc @tomaka

**Comment by larsbergstrom** [comment]
`Headless mode (OSMesa) is not supported on your platform.`, darn!

**Comment by tomaka** [comment]
Where do you get this error?
`OSMesa` is specific to Linux.

**Comment by larsbergstrom** [comment]
Hahaha! That's an error somebody put into our python driver code :-) I'll keep investigating and see what *actually* breaks in Rust code.

**Comment by larsbergstrom** [comment]
OMG, if I remove the python check, SERVO_HEADLESS just works! I'll do a timing run on our macpro shortly.

NIce work, @tomaka - headless on OSX just worked basically out of box :-)

**Comment by larsbergstrom** [comment]
@glennw @pcwalton 

I'm trying to run our CSS tests with headless Servo, but most of the tests fail with:
```
hread '<main>' panicked at 'Failed to compile shader: ERROR: 0:1: '' :  #version required and missing.
  │ ERROR: 0:8: 'varying' : syntax error: syntax error', /Users/larsberg/servo/.cargo/git/checkouts/rust-layers-c02709c0157e02cd/master/src/rendergl.rs:164
  │ stack backtrace:
  │    1:        0x104150888 - sys::backtrace::tracing::imp::write::h268482af15af0a80Y8t
  │    2:        0x1041529e5 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.42271
  │    3:        0x1041525e6 - panicking::default_handler::h97bc664f7e2bfeb5Ocy
  │    4:        0x10413e916 - sys_common::unwind::begin_unwind_inner::h98c0975c9b6b2fa0Zbt
  │    5:        0x10413edce - sys_common::unwind::begin_unwind_fmt::h3fbfbcdbf87c23935at
  │    6:        0x1035f566e - rendergl::ShaderProgram::compile_shader::hc4ed301c91d77f31VEa
  │    7:        0x1035f520a - rendergl::ShaderProgram::new::h7a05c4687bf57341jDa
  │    8:        0x1035f6591 - rendergl::TextureProgram::new::hbcfa31ef52f7f425wIa
  │    9:        0x1035f72bf - rendergl::TextureProgram::create_2d_program::hf045d7a9f5d87f29uMa
  │   10:        0x1035f79d9 - rendergl::RenderContext::new::hb70121bc7214d4e3zZa
  │   11:        0x10174a687 - compositor::IOCompositor<Window>::initialize_compositing::h3451773852088720166
  │   12:        0x1017491b9 - compositor::IOCompositor<Window>::handle_window_message::h6715051973487414355
  │   13:        0x10169343e - compositor::IOCompositor<Window>.CompositorEventListener::handle_events::h9158373627539836043
  │   14:        0x10175ed94 - Browser::handle_events::h9cd5198ed69c7366Mda
  │   15:        0x101659550 - main::h8517eb49d15e90fbNaa
  │   16:        0x104152212 - sys_common::unwind::try::try_fn::h9072539541001871658
  │   17:        0x10414fe2b - __rust_try
  │   18:        0x1041520c1 - rt::lang_start::hacfbaccc148e05fdU4x
  └   19:        0x101751329 - main
```

I'll keep poking at this, as headless testing looks like it would dramatically reduce our runtime on the builders, but if anything jumps out at you, it would help.

(you can test by setting `SERVO_HEADLESS=1` in the environment and then tweaking the headless support platform check in python/servo/build_commands.py to return `True`)

**Comment by glennw** [comment]
@larsbergstrom Depending on the version of OpenGL context you create, GLSL shaders may require you to specify the shader version with a #version preprocessor directive. So my guess would be that the headless option is creating a different version of opengl context than non-headless mode. Is that possible?

**Comment by larsbergstrom** [comment]
Hrm, I added `#version` directives, and walked my way back from OpenGL 3.0 / version 130 all the way to 110, but still get:

`thread '<main>' panicked at 'Failed to compile shader: ERROR: 0:1: '' :  version '110' is not supported`

This does not seem promising.

**Comment by larsbergstrom** [comment]
I find this really weird because we should be requesting OpenGL 3.2 successfully - I changed glutin's headless code like so:
```
[larsberg@lbergstrom glutin]$ git diff
diff --git a/src/api/cocoa/headless.rs b/src/api/cocoa/headless.rs
index 49b8bc3..84b747c 100644
--- a/src/api/cocoa/headless.rs
+++ b/src/api/cocoa/headless.rs
@@ -36,6 +36,8 @@ impl HeadlessContext {
                 NSOpenGLPFAAccelerated as u32,
                 NSOpenGLPFAAllowOfflineRenderers as u32,
                 NSOpenGLPFADoubleBuffer as u32,
+                NSOpenGLPFAOpenGLProfile as u32, // core profile with the version stated below
+                NSOpenGLProfileVersion3_2Core as u32,
```

**Comment by metajack** [comment]
@negge and I have been running into this with some of his glfw code on Mac. The Mac seems to be really picky about how this all gets set up, whereas linux doesn't seem to care at all. I suspect @tomaka or @glennw might know the correct incantations here.

**Comment by larsbergstrom** [comment]
That would be helpful! The headless offscreen code in glutin looks pretty much identical to that form Apple's dev docs, at least in terms of the offscreen framebuffer:
https://developer.apple.com/library/prerelease/mac/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/opengl_offscreen/opengl_offscreen.html#//apple_ref/doc/uid/TP40001987-CH403-SW2

I may have to try to reduce down my app to just glutin+layers and see if I can find something that lets me compile a shader.

**Comment by tomaka** [comment]
If you request the profile 3.2, then your shader won't compile.
If you don't put any `#version` header, it is assumed to be GLSL version 110, which isn't supported by OS X's core 3.2 profile.

I don't know how Servo works, but usually when you write an app you settle on a specific OpenGL version and 1) make sure you request the correct version at initialization. 2) put the correct `#version` header everywhere.

**Comment by frewsxcv** [comment]
This might have been fixed by #10935

**Comment by larsbergstrom** [comment]
Yeah, that's an alternative approach (using glutin's pseudo-headless mode instead of the true headless mode I was trying here) that should work at least as well.
